File: data/derived_cot/rq3_prompting/results/translation/qwen/cpp_to_py/code_output/PushBoxGame.py
class PushBoxGame:
    def __init__(self, map_data=None):
        if map_data is None:
            self.map = []
            self.player_row = 0
            self.player_col = 0
            self.targets = []
            self.boxes = []
            self.target_count = 0
            self._is_game_over = False
            self.init_game()
        else:
            self.map = map_data
            self.player_row = 0
            self.player_col = 0
            self.targets = []
            self.boxes = []
            self.target_count = 0
            self._is_game_over = False
            self.init_game()

    def init_game(self):
        for row_idx, row in enumerate(self.map):
            for col_idx, cell in enumerate(row):
                if cell == 'O':
                    self.player_row = row_idx
                    self.player_col = col_idx
                elif cell == 'G':
                    self.targets.append((row_idx, col_idx))
                    self.target_count += 1
                elif cell == 'X':
                    self.boxes.append((row_idx, col_idx))

    def gat_map(self):
        return self.map

    def is_game_over(self):
        return self._is_game_over

    def get_player_col(self):
        return self.player_col

    def get_player_row(self):
        return self.player_row

    def get_targets(self):
        return self.targets

    def get_boxes(self):
        return self.boxes

    def get_target_count(self):
        return self.target_count

    def check_win(self):
        box_on_target_count = 0
        for box in self.boxes:
            if box in self.targets:
                box_on_target_count += 1
        if box_on_target_count == self.target_count:
            self._is_game_over = True
        return self._is_game_over

    def move(self, direction):
        new_player_row = self.player_row
        new_player_col = self.player_col

        if direction == 'w':
            new_player_row -= 1
        elif direction == 's':
            new_player_row += 1
        elif direction == 'a':
            new_player_col -= 1
        elif direction == 'd':
            new_player_col += 1

        if self.map[new_player_row][new_player_col] == '#':
            return False

        if (new_player_row, new_player_col) in self.boxes:
            dx = new_player_row - self.player_row
            dy = new_player_col - self.player_col
            new_box_row = new_player_row + dx
            new_box_col = new_player_col + dy

            if self.map[new_box_row][new_box_col] == '#':
                return False

            box_to_move = (new_player_row, new_player_col)
            self.boxes.remove(box_to_move)
            self.boxes.append((new_box_row, new_box_col))

        self.player_row = new_player_row
        self.player_col = new_player_col

        self.check_win()
        return self._is_game_over

File: data/derived_cot/rq3_prompting/results/translation/qwen/py_to_cpp/code_output/PushBoxGame.cpp
#include <vector>
#include <string>
#include <utility> // for std::pair
#include <algorithm> // for std::find

class PushBoxGame {
private:
    std::vector<std::string> game_map;
    int player_row;
    int player_col;
    std::vector<std::pair<int, int>> targets;
    std::vector<std::pair<int, int>> boxes;
    int target_count;
    bool is_game_over;

    void init_game();

public:
    PushBoxGame(const std::vector<std::string>& map);
    bool check_win();
    bool move(char direction);
};

PushBoxGame::PushBoxGame(const std::vector<std::string>& map) {
    game_map = map;
    player_row = 0;
    player_col = 0;
    target_count = 0;
    is_game_over = false;
    init_game();
}

void PushBoxGame::init_game() {
    for (int row = 0; row < game_map.size(); row++) {
        for (int col = 0; col < game_map[row].size(); col++) {
            if (game_map[row][col] == 'O') {
                player_row = row;
                player_col = col;
            } else if (game_map[row][col] == 'G') {
                targets.push_back(std::make_pair(row, col));
                target_count++;
            } else if (game_map[row][col] == 'X') {
                boxes.push_back(std::make_pair(row, col));
            }
        }
    }
}

bool PushBoxGame::check_win() {
    int box_on_target_count = 0;
    for (const auto& box : boxes) {
        for (const auto& target : targets) {
            if (box.first == target.first && box.second == target.second) {
                box_on_target_count++;
                break;
            }
        }
    }
    if (box_on_target_count == target_count) {
        is_game_over = true;
    }
    return is_game_over;
}

bool PushBoxGame::move(char direction) {
    int new_player_row = player_row;
    int new_player_col = player_col;

    switch (direction) {
        case 'w': new_player_row--; break;
        case 's': new_player_row++; break;
        case 'a': new_player_col--; break;
        case 'd': new_player_col++; break;
        default: 
            return false;
    }

    if (new_player_row < 0 || new_player_row >= game_map.size() || 
        new_player_col < 0 || new_player_col >= game_map[new_player_row].size() ||
        game_map[new_player_row][new_player_col] == '#') {
        return false;
    }

    auto it = std::find(boxes.begin(), boxes.end(), std::make_pair(new_player_row, new_player_col));
    if (it != boxes.end()) {
        int new_box_row = new_player_row + (new_player_row - player_row);
        int new_box_col = new_player_col + (new_player_col - player_col);

        if (new_box_row < 0 || new_box_row >= game_map.size() || 
            new_box_col < 0 || new_box_col >= game_map[new_box_row].size() ||
            game_map[new_box_row][new_box_col] == '#') {
            return false;
        }

        boxes.erase(it);
        boxes.push_back(std::make_pair(new_box_row, new_box_col));
    }

    player_row = new_player_row;
    player_col = new_player_col;

    return check_win();
}